File: Algorithms/C++/Sorting/Quick_sort.cpp
//Input -
//8
//9 7 1 4 3 2 8 5
//Output
//1 2 3 4 5 7 8 9

//The Quick sort algo uses the concept of divide and conquer
//The time complexity of quick sort for below mentioned 3 cases are
//Best Case-0(n*logn)
//Worst Case-0(n^2)
//Average case-0(n*logn)

#include <iostream>
using namespace std;

void printArray(int arr[], int n);
void swap(int *, int *);

/*
1. After selecting an element as pivot, which is the first index of the array in our case, 
we divide the array for the first time.
2. In Quick sort this process is called partition where elements are positioned in such a way that
elements smaller than the pivot will be on left side and elements greater than the pivot will be on right side. 
3. The pivot element will be at its sorted position. After that we will again perform partition for the elements 
on left and right side of pivot by dividing them into subarrays.
*/
int partition(int arr[], int l, int h)
{
    //variable l stands for the starting position of the array,h stands for the
    //ending position of the array

    int pivot = arr[l];
    int i = l, j = h;

    while (i < j)
    {
        do
        {
            i++;
        } while (arr[i] <= pivot);

        do
        {
            j--;
        } while (arr[j] > pivot);

        if (i < j)
            swap(&arr[i], &arr[j]);
    }

    swap(&arr[l], &arr[j]);
    return j;
}

void quickSort(int arr[], int l, int h)
{

    if (l < h)
    {
        int p = partition(arr, l, h);
        quickSort(arr, l, p);
        quickSort(arr, p + 1, h);
    }
}

// The below code is for swapping the elements
void swap(int *i, int *j)
{
    int temp;
    temp = *i;
    *i = *j;
    *j = temp;
}

//The below function is used to print the array
void printArray(int arr[], int n)
{
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
    cout << "\n";
}

int main()
{

    int n;
    cout << "Enter no of elements";
    cin >> n;
    int arr[n];
    for (int i = 0; i < n; i++)
        cin >> arr[i];
    quickSort(arr, 0, n - 1);
    printArray(arr, n);
    return 0;
}

File: Algorithms/Java/SortingTechniques/QuickSort/QuickSort.java
package SortingTechniques.QuickSort;

import java.util.Scanner;

/************ Sorting Algorithm : Merge Sort **********************
 *
 * #GENERAL:
 * QuickSort is a Divide and Conquer algorithm. It picks an element as pivot and partitions the given array around the picked pivot.
 * There are many different versions of quickSort that pick pivot in different ways:
 * 1. Always pick first element as pivot.
 * 2. Always pick last element as pivot (implemented below)**
 * 3. Pick a random element as pivot.
 * 4. Pick median as pivot.
 *
 * The key process in quickSort is partition().
 * Target of partitions is, given an array and an element x of array as pivot,
 * put x at its correct position in sorted array and put all smaller elements (smaller than x) before x,
 * and put all greater elements (greater than x) after x. All this should be done in linear time.
 *
 * #TIME/SPACE COMPLEXITIES:
 * Best case Time Complexity: O(nLog(n))
 * Worst case Time Complexity: O(n^2) [^ => Exponentiation]
 * Average case Time Complexity: O(nLog(n))
 * Auxiliary Space: O(1)
 *
 * #RECURSION IDEA:
 * If low > high
 *      1. Decide the pivot and find the right place for the pivot
 *              :in our case we chose the last element to be the pivot
 *      2. Call quickSort for first half:
 *              Call quickSort(arr, low, p-1)
 *      3. Call quickSort for second half:
 *              Call quickSort(arr, p+1, high)
 *
 * #NOTE:
 * Please uncomment the lines in "/*" comment pattern in order to debug and print outputs to console.
 *******************************************************************/

public class QuickSort {

    // A utility function to swap two elements
    static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    // This function takes last element as pivot, places
    // the pivot element at its correct position in sorted
    // array, and places all smaller (smaller than pivot)
    // to left of pivot and all greater elements to right
    // of pivot
    static int partition(int[] arr, int low, int high) {

        // pivot
        int pivot = arr[high];
        /*
        System.out.println("Pivot: " + pivot);
         */

        // Index of smaller element and indicates the right position of pivot found so far
        int i = (low - 1);

        for (int j = low; j <= high - 1; j++) {
            // If current element is smaller than the pivot
            /*
            System.out.println("Comparing arr[j]: " + arr[j] + " with pivot: " + pivot);
             */
            if (arr[j] < pivot) {
                // Increment index of smaller element
                i++;
                /*
                System.out.println("--Going to swap " + arr[i] + " with " + arr[j]);
                 */
                swap(arr, i, j);
            }
        }
        /*
        System.out.print("Array before putting the pivot in right place --> ");
        printArrayFromSpecificStartAndEndPosition(arr, low, high);
        System.out.println("--Going to put Pivot in right place by swapping " + arr[i + 1] + " with " + arr[high]);
         */
        swap(arr, i + 1, high);
        return (i + 1);
    }

    // The main function that implements QuickSort
    // arr[] --> Array to be sorted,
    // low --> Starting index,
    // high --> Ending index

    static void quickSort(int[] arr, int low, int high) {
        /*
        System.out.print("Going to Sort --> ");
        printArrayFromSpecificStartAndEndPosition(arr, low, high);
         */
        if (low < high) {
            // pi is partitioning index, arr[p] is now at right place
            int pi = partition(arr, low, high);

            /*
            System.out.print("Array after doing the partition step: ---> ");
            printArrayFromSpecificStartAndEndPosition(arr, low, high);
            System.out.println();
            */
            // Separately sort elements before partition and after partition
            /*
            System.out.print("Left Array --> ");
            printArrayFromSpecificStartAndEndPosition(arr, low, pi - 1);
            System.out.print("Right Array --> ");
            printArrayFromSpecificStartAndEndPosition(arr, pi + 1, high);
             */
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    // A utility function to print array of size n from specific start and end points
    static void printArrayFromSpecificStartAndEndPosition(int[] array, int start, int end) {
        if (start <= end) {
            for (int i = start; i <= end; i++) {
                System.out.print(array[i] + " ");
            }
        } else {
            System.out.print("{}");
        }
        System.out.println();
    }

    // Function to print an array
    static void printArray(int[] arr) {
        for (int j : arr) System.out.print(j + " ");
        System.out.println();
    }

    // Driver Code
    public static void main(String[] args) {
        // Using Scanner for Getting Input from User
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter the number of elements in the array : ");
        int totalElements = scanner.nextInt();
        int[] array = new int[totalElements];

        System.out.println("Enter array elements:");
        for (int i = 0; i < totalElements; i++) {
            System.out.println("Enter " + i + " element: ");
            array[i] = scanner.nextInt();
        }

        System.out.println("Your input array :");
        printArray(array);

        quickSort(array, 0, array.length - 1);
        System.out.println("Sorted array");
        printArray(array);
    }
}